src/Microsoft.AspNet.Authentication.JwtBearer/Events/IJwtBearerEvents.cs

```diff
@@ -21,21 +21,21 @@ public interface IJwtBearerEvents
         /// <summary>
         /// Invoked when a protocol message is first received.
         /// </summary>
-        Task MessageReceived(MessageReceivedContext context);
+        Task ReceivingToken(ReceivingTokenContext context);
 
         /// <summary>
         /// Invoked with the security token that has been extracted from the protocol message.
         /// </summary>
-        Task SecurityTokenReceived(SecurityTokenReceivedContext context);
+        Task ReceivedToken(ReceivedTokenContext context);
 
         /// <summary>
         /// Invoked after the security token has passed validation and a ClaimsIdentity has been generated.
         /// </summary>
-        Task SecurityTokenValidated(SecurityTokenValidatedContext context);
+        Task ValidatedToken(ValidatedTokenContext context);
 
         /// <summary>
         /// Invoked to apply a challenge sent back to the caller.
         /// </summary>
-        Task ApplyChallenge(AuthenticationChallengeContext context);
+        Task SetAuthenticationHeader(SetAuthenticationHeaderContext context);
```

Comment:
I'm not totally convinced by this new name, which seems to imply we can only modify the headers from this event.

One of the most common use cases of this event is to write an error directly to the response body (+ the standard WWW-Authenticate header) to make authentication errors easier to diagnose.

(sample: http://graph.facebook.com/me)